### PR TITLE
Unsubscribe from chokidar file watching on stop request. Fixes #68.

### LINKED ISSUE #68: Unable to stop a (folder) stream
Hi,

When using SlpFolderStream+SlpRealtime, I'm unable to cancel/stop watching the folder or streams.

I've setup a bare reproduction repo: https://github.com/Milanzor/example-slp-realtime-not-closing

Mainly want to achieve this for testing, now I've checked the tests in this repo which also do things with SlpRealtime, but that uses RxSlpStream and manually pipes the slp through `pipeFileContents`. 

Perhaps I've overlooked something, in fact, I hope I have.

Thanks!

### PR DESCRIPTION
**Pull request by Milanzor**
Hi Vince,

Here's my PR that will fix the SlpFolderStream never ending, even when calling `.stop()` on it. Fixes #68.

What I've done:

## Added a new helper function `pipeFileContentsToFile`
This can be used to test the `SlpFolderStreams`, it it's basically a file-copy function that we can use to copy one of the `.slp` files to a temporary folder for a `SlpFolderStream` to pick up. 
It supports a `slow` parameter which uses [`brake`](https://www.npmjs.com/package/brake) to throttle the copy. This sort-of emulates a real `.slp` file being written by Dolphin and will allow us to test `SlpRealTime` events through `SlpFolderStreams` (those tests not in this PR though).

## Added SlpFolderStream tests
Only 2 basic tests, the first one only checks if the class can be instanced, the other does more:
- Sets up a new `SlpFolderStream`.
- Starts watching a temporary folder (`/slp/folder-stream`).
- Uses the new helper `pipeFileContentsToFile` to copy an existing `.slp` file to the temporary folder.
- Checks if the SlpFolderStream detects the new file and makes sure it is the file we expect it to be.
- Then stops the stream and checks if the chokidar watcher has been unset.
Note: I'm not super happy about this test, it works but does use some hacky setTimeouts, perhaps you can shed some light on it, the SlpFolderStream doesn't return a Promise in the setup/start function so there is no way of knowing it is ready and watching. I currently do not know how to make it do so. (RXJS is magic to me)

## Exposed the chokidar watcher + made the `.stop()` function close the watcher
Basically what this PR is for, `SlpFolderStream` now exposes the watcher using `.getWatcher()` and the `.stop()` method now stops and unsets the watcher.

Thanks!

### CODE REVIEW

**Comment by Milanzor** [comment]
Oh ye, I've also removed the `console.log` calls when the `SlpFolderStream` starts and when it finds a file. No other file in this project logs anything to the console, why should `SlpFolderStream` :)

File: .gitignore [hunk 1]
```diff
@@ -69,3 +69,6 @@ dist/
 *.old
 
 docs
+
+# Test files
+slp/folder-stream/*.slp
```
Comment (vinceau) [inline .gitignore]: I don't think we need to explicitly add this to the gitignore. I think it would be better for the test case to just clean up after itself once the test has completed.

File: src/stream/slpFolderStream.ts [hunk 2]
```diff
@@ -86,9 +88,15 @@ export class SlpFolderStream extends RxSlpStream {
       this.readStream.unpipe(this);
       this.readStream.done();
       this.readStream = null;
+      this.watcher?.close();
+      this.watcher = null;
```
Comment (vinceau) [inline src/stream/slpFolderStream.ts]:
```suggestion
     if (this.watcher) {
       this.watcher.close();
       this.watcher = null;
     }
```
I would prefer an explicit check here.

File: src/stream/slpFolderStream.ts [hunk 3]
```diff
@@ -86,9 +88,15 @@ export class SlpFolderStream extends RxSlpStream {
...
     }
   }
 
+  public getWatcher() {
+    return this.watcher;
+  }
+
```
Comment (vinceau) [inline src/stream/slpFolderStream.ts]: Do we really need this? Is it only exposed for the test?

File: src/utils/helpers.ts [hunk 4]
```diff
@@ -1,7 +1,8 @@
 import fs from "fs";
+const brake = require("brake");
```
Comment (vinceau) [inline src/utils/helpers.ts]: Imports in typescript should be:
```suggestion
import brake from "brake";
```
If you get type issues try run `yarn add -D @types/brake`.

File: src/utils/helpers.ts [hunk 5]
```diff
@@ -18,3 +19,37 @@ export const pipeFileContents = async (filename: string, destination: Writable,
...
+    let stopPipeStream = () => {
+      readStream.unpipe(writeStream).unpipe(brakeStream);
+      readStream.destroy();
+      writeStream.destroy();
+      brakeStream.destroy();
+    };
+
+    resolve({ readStream, writeStream, brakeStream, stopPipeStream });
+  });
+};
```
Comment (vinceau) [inline src/utils/helpers.ts]: I don't think the change to the helper function should be done in `utils/helpers.ts`. The whole brake stream thing seems to really only be applicable for the test so you can just make a copy of the function in the `spec.ts` file itself.

File: test/slpFolderStream.spec.ts [hunk 6]
```diff
@@ -0,0 +1,56 @@
+const path = require("path");
+const targetDirectory = path.resolve(__dirname, "..", "slp", "folder-stream");
+const fs = require("fs");
+
+import { pipeFileContentsToFile } from "../src";
+
+const { SlpFolderStream } = require("../src");
```
Comment (vinceau) [inline test/slpFolderStream.spec.ts]: Again refer to the typescript style of import as mentioned in the previous comment.

**Review by vinceau — CHANGES_REQUESTED** [review]
Hey! First up, thanks for looking into this and submitting a PR! I agree that the test could be set up a bit better. Specifically I don't think you really need `brake` at all. I would structure the test like this:
* start the folder stream
* copy a file over
* assert that the file was detected
* stop the folder stream
* copy another file over
* assert that the new file is not detected

You can use `sinon` in the test to assert that the number of event calls were a certain number. There should be able to find some examples of that in the codebase.

**Comment by Milanzor** [comment]
Thanks for looking at my PR, I'll submit changes shortly.

After going back and forth about `brake`, I agree we don't need it, I thought it would be a smart way to slowly test SlpRealtime, but there's no need to do that slowly, the `realtime.spec.ts` handles testing it just fine.

**Comment by Milanzor** [comment]
Stuck with waiting for SlpFolderStream to be ready:
Chokidar has the `ready` event, which will be fired when the watcher is fully ready, but SlpFolderStream is not setup to wait for that and I'm clueless how to fix that lol.

Could you point me in the right direction, if you have any idea? I can only think of fully changing SlpFolderStream to be async and don't resolve until the watcher and the rxjs subscriptions are all setup and ready.

**Comment by vinceau** [comment]
One way you could do that is just make the watcher `public` instead of `private`. Then you could just await the promise like this:
```js
await new Promise(resolve => {
  slpFolderStream.watcher.once('ready', resolve);
});
```

so then your test would look like:

```js
// psuedocode
const gameStartSpy = sinon.spy();
const slpFolderStream = new SlpFolderStream();

// Track the game start event
slpFolderStream.gameStart$.susbscribe(gameStartSpy);
slpFolderStream.start();

// wait til chokidar is finished watching
await new Promise(); // do the thing above

//copy the test file or whatever to the location

expect(gameStartSpy.callCount).toEqual(1);

// unsubscribe
slpFolderStream.stop();

// copy another file

// expect the call count to still be equal to 1.
```

Does that make sense?

I would also recommend looking at the other test files like this one: https://github.com/vinceau/slp-realtime/blob/master/test/combo.spec.ts to get a better idea on the testing setup. For example, I would imagine you could create the test directory in the `beforeAll` section, and in the `afterAll` section clean up/delete the files folders that you made.

**Comment by Milanzor** [comment]
Could you give it a spin? It's super janky and I'm not happy with it . Thing is, I feel it's related to rxjs handling things too late (asynchronously). The test runs, completes and cleans up the temp files and afterwards, rxjs tries to run it's subscriber events while the slp files don't exist anymore.

Perhaps I could drop the tests, remove this PR and create a new PR that simply exposes the watcher, run `.close` on it in the `SlpFolderStream.stop()` function and be done with it? I also don't wanna be too much of a burden on your time simply because I don't understand rxjs well enough.

Thanks for your time anyway!

**Comment by Milanzor** [comment]
By the way, if/when the test properly runs, I'll add the sinon checks. These last commits were just trying to make it not fail/stay-alive

**Review by vinceau — APPROVED** [review]

**Comment by vinceau** [comment]
You're right. There were a few issues with the testing setup due to how things were being run asynchronously. Adding some artificial delay helps with that. I've fixed up the test for you. Thanks for your help! Let me know if this change fixes your issues.
